homeassistant/components/sensor/bmw_connected_drive.py

```diff
@@ -0,0 +1,98 @@
+"""
+Reads vehicle status from BMW connected drive portal.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.bmw_connected_drive/
+"""
+import logging
+import asyncio
+
+from homeassistant.components.bmw_connected_drive import DOMAIN as BMW_DOMAIN
+from homeassistant.helpers.entity import Entity
+
+DEPENDENCIES = ['bmw_connected_drive']
+
+_LOGGER = logging.getLogger(__name__)
+
+LENGTH_ATTRIBUTES = [
+    'remaining_range_fuel',
+    'mileage',
+    ]
+
+VALID_ATTRIBUTES = LENGTH_ATTRIBUTES + [
+    'remaining_fuel',
+]
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the BMW sensors."""
+    accounts = hass.data[BMW_DOMAIN]
+    _LOGGER.debug('Found BMW accounts: %s',
+                  ', '.join([a.name for a in accounts]))
+    devices = []
+    for account in accounts:
+        for vehicle in account.account.vehicles:
+            for sensor in VALID_ATTRIBUTES:
+                device = BMWConnectedDriveSensor(account, vehicle, sensor)
+                devices.append(device)
+    return add_devices(devices)
+
+
+class BMWConnectedDriveSensor(Entity):
+    """Representation of a BMW vehicle sensor."""
+
+    def __init__(self, account, vehicle, attribute: str):
+        """Constructor."""
+        self._vehicle = vehicle
+        self._account = account
+        self._attribute = attribute
+        self._state = None
+        self._unit_of_measurement = None
+        self._name = '{} {}'.format(self._vehicle.modelName, self._attribute)
+
+    @property
+    def should_poll(self) -> bool:
+        """Data update is triggered from BMWConnectedDriveEntity."""
+        return False
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor.
+
+        The return type of this call depends on the attribute that
+        is configured.
+        """
+        return self._state
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Get the unit of measurement."""
+        return self._unit_of_measurement
+
+    def update(self) -> None:
+        """Read new state data from the library."""
+        vehicle_state = self._vehicle.state
+        self._state = getattr(vehicle_state, self._attribute)
+
+        if self._attribute in LENGTH_ATTRIBUTES:
+            self._unit_of_measurement = vehicle_state.unit_of_length
+        elif self._attribute == 'remaining_fuel':
+            self._unit_of_measurement = vehicle_state.unit_of_volume
+        else:
+            self._unit_of_measurement = None
+
+        self.schedule_update_ha_state()
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Add callback after being added to hass.
+
+        Show latest data after startup.
+        """
+        self._account.add_update_listener(self.update)
+        self.update()
```

Comment:
Remove this.

Reply:
What should I remove?

The self.update? I need this so the state is available right after the Start of hass. If I want to trigger this from the component, I would have to find out when all of the setup operations were completed.

Reply:
Yes remove this line here. See my comment above about how to add devices so that update will be called when they are added.

Reply:
Ok, so an alternative if you really want to keep the logging with entity_id in `update` is to schedule a call to `update` here. In that case, use `yield from self.hass.async_add_job(self.update)`.

Reply:
When I call `add_devices(devices, True)` from `setup_platform` even without me logging the entity_id, I get this error:
```
2018-02-19 16:55:32 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/home/kuehnelc/git/home-assistant/homeassistant/helpers/entity.py", line 196, in async_update_ha_state
    "No entity id specified for entity {}".format(self.name))
homeassistant.exceptions.NoEntitySpecifiedError: No entity id specified for entity 530i xDrive remaining_range_fuel
```
I'm not sure why that is.

Your version with `yield from self.hass.async_add_job(self.update)` works just as fine as `self.update()`. So I'll take your version. Not sure why you want update do be called via `async_add_job`, it's only doing a bit of string processing. There is no IO in the update function. But as it works, I'm fine with that :smile: ...

Reply:
The `update` method calls `schedule_update_ha_state` which needs to be done from a sync context.

Reply:
Regarding error, yeah sorry, the `entity_id` is also required when scheduling a state update, ie in `schedule_update_ha_state`, so the my suggestion to call `add_devices` like that was totally wrong. The logic about this changed recently and I hadn't really understood all the consequences until now.